### PR TITLE
Refactor Tcache to use RzHeapBin

### LINKED ISSUE #1088: Fix Heap commands for different versions of glibc
_Originally posted by @merrychap in https://github.com/rizinorg/rizin/issues/810#issuecomment-802013632_
### Testcase and environment

I tried to run it against several libcs:
- `libc6_2.24-3ubuntu1_amd64` - no tcache
- `libc6_2.27-3ubuntu1.3_amd64` - + tcache
- `libc6_2.29-0ubuntu2_amd64` - + tcache key
- `libc6_2.33-0ubuntu2_amd64` - + safe-linking

Here is my testcase:

```cpp
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <unistd.h>

int
main(int argc, char *argv[])
{
	int var = 0;

	uint64_t *p[100] = {0};

// #ifdef TCACHE
	
// #else
	p[0]  = malloc(0x48);
	p[1]  = malloc(0x48);
	p[2]  = malloc(0x48);
	p[3]  = malloc(0x68);
	p[4]  = malloc(0x508);
	p[5]  = malloc(0x68);
	p[6]  = malloc(0x508);
	p[7]  = malloc(0x68);
	p[8]  = malloc(0x428);
	p[9]  = malloc(0x28);
	p[10] = malloc(0x428);
	p[11] = malloc(0x28); // border

	free(p[0]); // fastbin[0x50] / tcache[0x50] 
	free(p[1]); // fastbin[0x50] / tcache[0x50]
	free(p[2]); // fastbin[0x50] / tcache[0x50]
	free(p[3]); // fastbin[0x70] / tcache[0x70]
	free(p[4]); // unsortedbin
	free(p[5]); // fastbin[0x70] / tcache[0x70]
	free(p[6]); // unsortedbin
	free(p[7]); // fastbin[0x70] / tcache[0x70]
	free(p[8]); // unsortedbin
	free(p[9]); // fastbin[0x30] / tcache[0x30]
	free(p[10]); // unsortedbin

	// examine fastbin / tcache and unsorted
	read(0, &var, 1);

	// malloc fastbin / tcache chunks back, so no consolidation will happen
	for (int i = 0; i < 3; i++) {
		malloc(0x48);
		malloc(0x68);
	}
	malloc(0x28);

	p[12] = malloc(0x3f8); // p[10] now 0x30
	p[13] = malloc(0x3f8); // p[10] in smallbin, p[8] now 0x30
	p[14] = malloc(0x608); // p[8] in smallbin, p[6] in largebin, p[4] in largebin

	// examine small and large
	read(0, &var, 1);

// #endif

	getchar();

	return 0;
}
```

Compile command: `gcc ./rizin_test.c -o rizin_test`
GCC version: `gcc (GCC) 10.2.0`
Run command: `LD_PRELOAD=~/libc/libc.so.6 ~/libc/ld-linux-x86-64.so.2 ./rizin_test`
rizin attach: ```./build/binrz/rizin/rizin -d `pgrep ld-linux-x86-64.so.2` ```

For the run command, I used `libc` and `ld` of the corresponding version. So the actual run command could be `LD_PRELOAD=~/libs/libc6_2.33-0ubuntu2_amd64/libc-2.33.so ~/libs/libc6_2.33-0ubuntu2_amd64/ld-2.33.so ./rizin_test`

### libc6_2.24-3ubuntu1_amd64

**Overall**: Looks like there are no visible bugs

### libc6_2.27-3ubuntu1.3_amd64

**Overall**: Most of the functionality doesn't work. I wasn't able to test all of the possible tcache functionality (so my testcase isn't completed properly) due to some logic errors listed below.

0. **Examine chunks using dmh**

```bash
[0x7fe1d0d00151]> dmh
Arena @0x7fe1d0fdbc40
Chunk(addr=0x0, size=0xfffffffffffffff8, flags=NON_MAIN_ARENA,IS_MMAPPED,PREV_INUSE)[top]
```

Incorrect address, size and flags. `rizin` prints no chunks, but there should be these chunks (from the previous examples):

```bash
[0x7fef23555c70]> dmh
Arena @0x7fef2381fb00
Chunk(addr=0x55555641a000, size=0x50, [skipped]
Chunk(addr=0x55555641a050, size=0x50, [skipped]
Chunk(addr=0x55555641a0a0, size=0x50, [skipped]
Chunk(addr=0x55555641a0f0, size=0x70, [skipped]
Chunk(addr=0x55555641a160, size=0x510, [skipped]
Chunk(addr=0x55555641a670, size=0x70, [skipped]
Chunk(addr=0x55555641a6e0, size=0x510, [skipped]
Chunk(addr=0x55555641abf0, size=0x70, [skipped]
Chunk(addr=0x55555641ac60, size=0x400, [skipped]
Chunk(addr=0x55555641b060, size=0x30, [skipped]
Chunk(addr=0x55555641b090, size=0x30, [skipped]
Chunk(addr=0x55555641b0c0, size=0x400, [skipped]
Chunk(addr=0x55555641b4c0, size=0x30, [skipped]
Chunk(addr=0x55555641b4f0, size=0x30, [skipped]
Chunk(addr=0x55555641b520, size=0x610, [skipped]
Chunk(addr=0x55555641bb30, size=0x1f4d0, [skipped]
```

The most significant difference from the previous example is tcache. So, maybe the problem lies somewhere in its processing.

1. **Examine chunks using dmhd**

```bash
[0x7fe1d0d00151]> dmhd

^C^C^C^Z
```

Results in an infinite loop with no possibility to cancel the operation using `Ctrl+C`.

2. **Examine tcache chunks**

```bash
[0x7fb7e03bb151]> dmht
[0x7fb7e03bb151]>
```

Results in no tcache chunks output, but tcache chunks exist. Look at `gef` output:

```bash
Tcachebins[idx=1, size=0x30] count=1  ←  Chunk(addr=0x555555cba2f0, size=0x30, flags=) 
Tcachebins[idx=3, size=0x50] count=3  ←  Chunk(addr=0x555555cb9300, size=0x50, flags=PREV_INUSE)  ←  Chunk(addr=0x555555cb92b0, size=0x50, flags=PREV_INUSE)  ←  Chunk(addr=0x555555cb9260, size=0x50, flags=PREV_INUSE) 
Tcachebins[idx=5, size=0x70] count=3  ←  Chunk(addr=0x555555cb9e50, size=0x70, flags=)  ←  Chunk(addr=0x555555cb98d0, size=0x70, flags=)  ←  Chunk(addr=0x555555cb9350, size=0x70, flags=PREV_INUSE)
```

This is what `rizin` supposed to print. Maybe the source of this bug is the same as the above examples - something with tcache processing (but I can be wrong of course).

3. **Examine unsorted chunks**

```bash
[0x7fb7e03bb151]> dmhd unsorted
^C^C^C^C^C^C^C^C^C^C^Z
```

Results in an infinite loop with no possibility to cancel the operation using `Ctrl+C`.

### libc6_2.29-0ubuntu2_amd64

**Overall**: Same problems as in `libc6_2.27-3ubuntu1.3_amd64`

### libc6_2.33-0ubuntu2_amd64

**Overall**: Same problems as in `libc6_2.27-3ubuntu1.3_amd64`

### PR DESCRIPTION
**Pull request by MalhotraPulak**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
Goals:
- Refactor the Tcache parsing code to use RzHeapBin so it can be easily integrated with Cutter. 
- Fix #1088 

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by XVilka** [comment]
There are some conflicts, please rebase on top of the latest `dev`.

File: librz/core/linux_heap_glibc.c [hunk 1]
```diff
@@ -2508,4 +2536,31 @@ RZ_API MallocState *GH(rz_heap_get_arena)(RzCore *core, GHT m_state) {
 		return NULL;
 	}
 	return main_arena;
+}
+
+RZ_API bool GH(rz_heap_write_heap_chunk)(RzCore *core, RzHeapChunkSimple *chunk_simple) {
```
Comment (XVilka) [inline librz/core/linux_heap_glibc.c:2550]: Please add Doxygen here since it's an API function

Comment (XVilka) [reply librz/core/linux_heap_glibc.c:2550]: Also it makes sense to check if `chunk_simple` is NULL

**Review by XVilka — CHANGES_REQUESTED** [review]

**Review by yossizap — APPROVED** [review]

**Comment by XVilka** [comment]
@PulakIIIT is this still a draft?

**Comment by MalhotraPulak** [comment]
> @PulakIIIT is this still a draft?

@XVilka there is a couple of bug fixes I am working if I am able to resolve them Ill push to this branch today, this PR should be merged by today or tomorrow

**Review by XVilka — APPROVED** [review]

File: librz/core/linux_heap_glibc.c [hunk 2]
```diff
@@ -125,6 +128,13 @@ static bool GH(is_tcache)(RzCore *core) {
 		return tcv != 0;
 	}
 	if (fp) {
+
+		// In case there is string `libc-` in path actual libc go to last occurrence of `libc-`
+		while (strstr(fp + 1, "libc-") != NULL) {
+			eprintf("%s\n", fp + 1);
```
Comment (wargio) [inline librz/core/linux_heap_glibc.c]: is this a debug string?

Comment (MalhotraPulak) [reply librz/core/linux_heap_glibc.c]: Nice catch! the eprintf was for debugging

**Pull request merged**